# Gather

#### Download the file 'twitter-archive-enhanced.csv' from the link [twitter_archive_enhanced.csv](https://d17h27t6h515a5.cloudfront.net/topher/2017/August/59a4e958_twitter-archive-enhanced/twitter-archive-enhanced.csv)

In [25]:
import pandas as pd

In [26]:
# Load data frame with twitter-archive-enhanced.csv
tweet_archive_data = pd.read_csv('twitter-archive-enhanced.csv')
tweet_archive_data.head(2)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None

#### Download the image_predictions.tsv programmatically from this [url](https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv)

In [27]:
import requests
import os

In [4]:
# download the file from the below url
url ='https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)
response

<Response [200]>

In [5]:
with open(url.split('/')[-1], mode='wb') as file:

    file.write(response.content)

In [6]:
os.listdir(os.curdir)

['.DS_Store',
 '.git',
 '.ipynb_checkpoints',
 'act_report.html',
 'act_report.ipynb',
 'image-predictions.tsv',
 'README.md',
 'tweet_archive_data_clean.csv',
 'tweet_count_data_clean.csv',
 'tweet_image_data_clean.csv',
 'tweet_json.txt',
 'twitter-archive-enhanced.csv',
 'twitter_archive_master.csv',
 'wrangle_act.html',
 'wrangle_act.ipynb']

In [28]:
# Load dataframe with image-predictions.tsv 
tweet_image_data = pd.read_csv('image-predictions.tsv', sep = '\t')
tweet_image_data.head(2)

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True

#### Gather tweet and retweet counts  for the tweet ids in the archive dataset

In [8]:
# Query Twitter API for each tweet in the Twitter archive and save JSON in a text file
# Keys are hidden to comply with Twitter's API terms and conditions
# Two rerun this replace the hidden with your own keys after subscribing to twitter developer account
import tweepy

consumer_key = 'Hidden-Key'
consumer_secret = 'Hidden-Key'
access_token = 'Hidden-Key'
access_secret = 'Hidden-Key'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True)

In [29]:
import json
from timeit import default_timer as timer

In [30]:
# Tweet ids in the twitter-archive-enhanced.csv file
tweet_ids = tweet_archive_data.tweet_id.values
len(tweet_ids)

2356

In [11]:
# Query Twitter's API for JSON data for each tweet ID in the Twitter archive
count = 0
fails_dict = {}
start = timer()
# Save each tweet's returned JSON as a new line in a .txt file
with open('tweet_json.txt', 'w') as outfile:
    # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
    for tweet_id in tweet_ids:
        count += 1
        # print(str(count) + ": " + str(tweet_id))
        try:
            tweet = api.get_status(tweet_id, tweet_mode='extended')
            # print("Success")
            json.dump(tweet._json, outfile)
            outfile.write('\n')
        except tweepy.TweepError as e:
            # print("Fail")
            fails_dict[tweet_id] = e
            pass
end = timer()
print(end - start)
print(fails_dict)

1891.9851507429994
{888202515573088257: TweepError([{'code': 144, 'message': 'No status found with that ID.'}],), 873697596434513921: TweepError([{'code': 144, 'message': 'No status found with that ID.'}],), 872668790621863937: TweepError([{'code': 144, 'message': 'No status found with that ID.'}],), 869988702071779329: TweepError([{'code': 144, 'message': 'No status found with that ID.'}],), 866816280283807744: TweepError([{'code': 144, 'message': 'No status found with that ID.'}],), 861769973181624320: TweepError([{'code': 144, 'message': 'No status found with that ID.'}],), 845459076796616705: TweepError([{'code': 144, 'message': 'No status found with that ID.'}],), 842892208864923648: TweepError([{'code': 144, 'message': 'No status found with that ID.'}],), 837012587749474308: TweepError([{'code': 144, 'message': 'No status found with that ID.'}],), 827228250799742977: TweepError([{'code': 144, 'message': 'No status found with that ID.'}],), 812747805718642688: TweepError([{'code':

In [12]:
# how many twitter ids the data retreival failed
len(fails_dict)

16

In [31]:
# read json data from tweet_json.txt and load into a dataframe
tweet_count_list = []
with open('tweet_json.txt') as json_file:
    for line in json_file:
        json_data = json.loads(line)
        tweet_id = json_data['id']
        retweet_count = json_data['retweet_count']
        favorite_count = json_data['favorite_count']
        tweet_count_list.append({'tweet_id': tweet_id,
                        'retweet_count': int(retweet_count),
                        'favorite_count': int(favorite_count)
                        })



tweet_count_data = pd.DataFrame(tweet_count_list, columns = ['tweet_id', 'retweet_count', 'favorite_count'])
        

In [32]:
tweet_count_data.head(2)

tweet_id  retweet_count  favorite_count
0  892420643555336193           8257           37848
1  892177421306343426           6101           32523

# Assess

### Visual Assessment
#### For tweet_archive_data

In [33]:
tweet_archive_data

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [34]:
tweet_archive_data.head(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [35]:
tweet_archive_data.tail(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
2351  https://twitter.com/dog_rates/status/666049248...                 5   
2352  https://twitter.com/dog_rates/status/666044226...                 6   
2353  https://twitter.com/dog_rates/status/666033412...                 9   
2354  https://twitter.com/dog_rates/status/666029285...                 7   
2355  https://twitter.com/dog_rates/status/666020888...                 8   

      rating_denominator  name doggo floofer pupper puppo  
2351                  10  None  None    None   None  None  
2352                  10     a  None    None   None  None  
2353                  10     a  None    None   None  None  
2354                  10     a  None    None   None  None  
2355                  10  None  None    None   None  None

In [37]:
tweet_archive_data.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
458   817908911860748288                    NaN                  NaN   
2020  672095186491711488                    NaN                  NaN   
1664  682788441537560576                    NaN                  NaN   
916   756998049151549440                    NaN                  NaN   
1343  704761120771465216                    NaN                  NaN   

                      timestamp  \
458   2017-01-08 01:40:55 +0000   
2020  2015-12-02 16:49:14 +0000   
1664  2016-01-01 05:00:24 +0000   
916   2016-07-23 23:42:53 +0000   
1343  2016-03-01 20:11:59 +0000   

                                                 source  \
458   <a href="http://twitter.com/download/iphone" r...   
2020  <a href="http://twitter.com/download/iphone" r...   
1664  <a href="http://twitter.com/download/iphone" r...   
916   <a href="http://twitter.com/download/iphone" r...   
1343  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
458   Looks like he went cross-eyed trying way too h...                  NaN   
2020  This is Tuco. That's the toast that killed his...                  NaN   
1664  Happy New Year from your fav holiday squad! 🎉 ...                  NaN   
916   This is Oliver. He's an English Creamschnitzel...                  NaN   
1343  This pupper killed this great white in an epic...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
458                        NaN                        NaN   
2020                       NaN                        NaN   
1664                       NaN                        NaN   
916                        NaN                        NaN   
1343                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
458   https://twitter.com/micahgrimes/status/8179020...                12   
2020  https://twitter.com/dog_rates/status/672095186...                 9   
1664  https://twitter.com/dog_rates/status/682788441...                12   
916   https://twitter.com/dog_rates/status/756998049...                11   
1343  https://twitter.com/dog_rates/status/704761120...                13   

      rating_denominator    name doggo floofer  pupper puppo  
458                   10    None  None    None    None  None  
2020                  10    Tuco  None    None    None  None  
1664                  10    None  None    None  pupper  None  
916                   10  Oliver  None    None    None  None  
1343                  10    None  None    None  pupper  None

##### Upon just visual observation we can see that
##### 1. we have 17 observations in this data set
##### 2. There are NaN values in in_reply_to_status_id , in_reply_to_user_id , retweeted_status_id, retweeted_status_user_id and retweeted_status_timestamp columns. 
##### 3. Column "name" seem to represent "None" values and incorrect values like " a,the, an, etc"
##### 4. Some records have a denominator value different from 10
##### 5. Last four columns seem to be holding dog breed type and this structurally untidy. 

### Programmatic Assessment
#### For tweet_archive_data

In [10]:
tweet_archive_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

##### tweet_archive_data has 17 columns and 2356 observations, tweet_id seems to be the unique Id
##### in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id , retweeted_status_user_id, retweeted_status_timestamp columns has very few non null values.

##### timestamp column is not represented as date.

##### Let's see if there are any duplicate records based on tweet_id

In [15]:
tweet_archive_data[tweet_archive_data.tweet_id.duplicated()]

Empty DataFrame
Columns: [tweet_id, in_reply_to_status_id, in_reply_to_user_id, timestamp, source, text, retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, expanded_urls, rating_numerator, rating_denominator, name, doggo, floofer, pupper, puppo]
Index: []

#### No duplicates in the table for tweet_id

In [38]:
tweet_archive_data.tweet_id.nunique()

2356

##### tweet_id is the unique id and there are 2356 unique tweet_ids for 2356 observations in the tweet_archive_data 


##### Let's take a closer look at columns where we have majority NaN Values

In [19]:
tweet_archive_data.isnull().sum()

tweet_id                         0
in_reply_to_status_id         2278
in_reply_to_user_id           2278
timestamp                        0
source                           0
text                             0
retweeted_status_id           2175
retweeted_status_user_id      2175
retweeted_status_timestamp    2175
expanded_urls                   59
rating_numerator                 0
rating_denominator               0
name                             0
doggo                            0
floofer                          0
pupper                           0
puppo                            0
dtype: int64

##### in_reply_to_status_id  has 2278 NaN
##### in_reply_to_user_id has 2278 NaN
##### retweeted_status_id has 2175 NaN
##### retweeted_status_user_id 2175 NaN
##### retweeted_status_timestamp 2175 NaN
##### Since these columns have majority values as Null these columns and we currently do  not have a way to find the missing values we will be  deleting these columns from the data frame.


##### Let's take a closer look at the values populated in the "name" column

In [21]:
tweet_archive_data.name.sample(2)

3       Darla
1132     None
Name: name, dtype: object

In [22]:
tweet_archive_data.name.value_counts()

None            745
a                55
Charlie          12
Lucy             11
Cooper           11
Oliver           11
Tucker           10
Lola             10
Penny            10
Bo                9
Winston           9
the               8
Sadie             8
Daisy             7
Bailey            7
Toby              7
Buddy             7
an                7
Jack              6
Jax               6
Oscar             6
Leo               6
Koda              6
Dave              6
Scout             6
Rusty             6
Milo              6
Bella             6
Stanley           6
Oakley            5
               ... 
Cecil             1
Yukon             1
Ziva              1
Sonny             1
Enchilada         1
Burt              1
Asher             1
Jazz              1
Strudel           1
Taco              1
Lance             1
Brutus            1
Alexanderson      1
Cannon            1
Pupcasso          1
Farfle            1
Boston            1
Kallie            1
Ike               1


##### Some of the values in the name column are "None","the","such" ,"a" and "an". These dont look like real names of the dogs. Majority of the rows ( 745) have None populated. Does this mean these rows dont have a Dog name available and data is entered as None.

##### All the valid values for the name are starting with Uppercase letter.
##### All the invalid values are starting with lowercase
##### let's take a closer look at the invalid name values and their counts

In [23]:
tweet_archive_data.loc[tweet_archive_data['name'].str.islower()].name.value_counts()

a               55
the              8
an               7
very             5
quite            4
just             4
one              4
getting          2
mad              2
actually         2
not              2
unacceptable     1
light            1
all              1
such             1
officially       1
life             1
space            1
incredibly       1
this             1
by               1
my               1
infuriating      1
old              1
his              1
Name: name, dtype: int64

In [24]:
tweet_archive_data.loc[tweet_archive_data['name'].str.islower()].name.count()

109

##### There are 109 rows that have invalid values for name
##### There are 745 rows that have name value as None


##### Let's  see where the tweets are originated from

In [25]:
tweet_archive_data.source.value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                          91
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                       33
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>      11
Name: source, dtype: int64

##### Looks like the majority of the tweets are originated from an iphone

##### Let's look at the tweet text data

In [46]:
tweet_archive_data.text.nunique()

2356

In [48]:
tweet_archive_data.text.value_counts

<bound method IndexOpsMixin.value_counts of 0       This is Phineas. He's a mystical boy. Only eve...
1       This is Tilly. She's just checking pup on you....
2       This is Archie. He is a rare Norwegian Pouncin...
3       This is Darla. She commenced a snooze mid meal...
4       This is Franklin. He would like you to stop ca...
5       Here we have a majestic great white breaching ...
6       Meet Jax. He enjoys ice cream so much he gets ...
7       When you watch your owner call another dog a g...
8       This is Zoey. She doesn't want to be one of th...
9       This is Cassie. She is a college pup. Studying...
10      This is Koda. He is a South Australian decksha...
11      This is Bruno. He is a service shark. Only get...
12      Here's a puppo that seems to be on the fence a...
13      This is Ted. He does his best. Sometimes that'...
14      This is Stuart. He's sporting his favorite fan...
15      This is Oliver. You're witnessing one of his m...
16      This is Jim. He foun

In [51]:
tweet_archive_data[['name','text']]

name                                               text
0      Phineas  This is Phineas. He's a mystical boy. Only eve...
1        Tilly  This is Tilly. She's just checking pup on you....
2       Archie  This is Archie. He is a rare Norwegian Pouncin...
3        Darla  This is Darla. She commenced a snooze mid meal...
4     Franklin  This is Franklin. He would like you to stop ca...
5         None  Here we have a majestic great white breaching ...
6          Jax  Meet Jax. He enjoys ice cream so much he gets ...
7         None  When you watch your owner call another dog a g...
8         Zoey  This is Zoey. She doesn't want to be one of th...
9       Cassie  This is Cassie. She is a college pup. Studying...
10        Koda  This is Koda. He is a South Australian decksha...
11       Bruno  This is Bruno. He is a service shark. Only get...
12        None  Here's a puppo that seems to be on the fence a...
13         Ted  This is Ted. He does his best. Sometimes that'...
14      Stuart  This is Stuart. He's sporting his favorite fan...
15      Oliver  This is Oliver. You're witnessing one of his m...
16         Jim  This is Jim. He found a fren. Taught him how t...
17        Zeke  This is Zeke. He has a new stick. Very proud o...
18     Ralphus  This is Ralphus. He's powering up. Attempting ...
19      Canela  RT @dog_rates: This is Canela. She attempted s...
20      Gerald  This is Gerald. He was just told he didn't get...
21     Jeffrey  This is Jeffrey. He has a monopoly on the pool...
22        such  I've yet to rate a Venezuelan Hover Wiener. Th...
23      Canela  This is Canela. She attempted some fancy porch...
24        None  You may not have known you needed to see this ...
25        None  This... is a Jubilant Antarctic House Bear. We...
26        Maya  This is Maya. She's very shy. Rarely leaves he...
27      Mingus  This is Mingus. He's a wonderful father to his...
28       Derek  This is Derek. He's late for a dog meeting. 13...
29      Roscoe  This is Roscoe. Another pupper fallen victim t...
...        ...                                                ...
2326     quite  This is quite the dog. Gets really excited whe...
2327         a  This is a southern Vesuvius bumblegruff. Can d...
2328      None  Oh goodness. A super rare northeast Qdoba kang...
2329      None  Those are sunglasses and a jean jacket. 11/10 ...
2330      None  Unique dog here. Very small. Lives in containe...
2331      None  Here we have a mixed Asiago from the Galápagos...
2332      None  Look at this jokester thinking seat belt laws ...
2333        an  This is an extremely rare horned Parthenon. No...
2334         a  This is a funny dog. Weird toes. Won't come do...
2335        an  This is an Albanian 3 1/2 legged  Episcopalian...
2336      None     Can take selfies 11/10 https://t.co/ws2AMaNwPW
2337      None  Very concerned about fellow dog trapped in com...
2338      None  Not familiar with this breed. No tail (weird)....
2339      None  Oh my. Here you are seeing an Adobe Setter giv...
2340      None  Can stand on stump for what seems like a while...
2341      None  This appears to be a Mongolian Presbyterian mi...
2342      None  Here we have a well-established sunblockerspan...
2343      None  Let's hope this flight isn't Malaysian (lol). ...
2344      None  Here we have a northern speckled Rhododendron....
2345       the  This is the happiest dog you will ever see. Ve...
2346       the  Here is the Rand Paul of retrievers folks! He'...
2347         a  My oh my. This is a rare blond Canadian terrie...
2348         a  Here is a Siberian heavily armored polar bear ...
2349        an  This is an odd dog. Hard on the outside but lo...
2350         a  This is a truly beautiful English Wilson Staff...
2351      None  Here we have a 1949 1st generation vulpix. Enj...
2352         a  This is a purebred Piers Morgan. Loves to Netf...
2353         a  Here is a very happy pup. Big fan of well-main...
2354         a  This is a western brown Mitsubishi

##### It looks like name is extracted from the text value.

#### let's see sample records for the text value where the name is populated a None

In [55]:
tweet_archive_data.query('name=="None"')[['text']].sample(5)

text
1949  When you're having a blast and remember tomorr...
839   I don't know any of the backstory behind this ...
923   Here's a frustrated pupper attempting to escap...
1237  Reminder that we made our first set of sticker...
1445    Oh my god 10/10 for every little hot dog pupper

##### It looks like the text does not contain the name of the dog, hence the parser couldnt not pick up any value for the dog name

##### Let's take a look at expanded_urls data

In [59]:
tweet_archive_data.expanded_urls.sample(5)

1522    https://twitter.com/dog_rates/status/690649993...
378     https://twitter.com/dog_rates/status/827933404...
744     https://twitter.com/dog_rates/status/780192070...
591     https://twitter.com/dog_rates/status/799063482...
922     https://twitter.com/dog_rates/status/756275833...
Name: expanded_urls, dtype: object

In [58]:
tweet_archive_data.expanded_urls.nunique()

2218

##### There are 2218 unique expanded urls!

In [60]:
tweet_archive_data.expanded_urls.value_counts()

https://twitter.com/dog_rates/status/694669722378485760/photo/1,https://twitter.com/dog_rates/status/694669722378485760/photo/1                                                                                                                                                                                                                                                                    2
https://twitter.com/dog_rates/status/698195409219559425/photo/1                                                                                                                                                                                                                                                                                                                                    2
https://twitter.com/dog_rates/status/667182792070062081/photo/1                                                                                                                                                               

#### Let's look at the rating_denominator and rating_numerator values

In [29]:
tweet_archive_data.rating_denominator.value_counts()

10     2333
11        3
50        3
80        2
20        2
2         1
16        1
40        1
70        1
15        1
90        1
110       1
120       1
130       1
150       1
170       1
7         1
0         1
Name: rating_denominator, dtype: int64

##### Majority of the data , 2333 rows have a valid denominator of 10

In [30]:
tweet_archive_data.query("rating_denominator != 10")

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
313   835246439529840640           8.352460e+17         2.625958e+07   
342   832088576586297345           8.320875e+17         3.058208e+07   
433   820690176645140481                    NaN                  NaN   
516   810984652412424192                    NaN                  NaN   
784   775096608509886464                    NaN                  NaN   
902   758467244762497024                    NaN                  NaN   
1068  740373189193256964                    NaN                  NaN   
1120  731156023742988288                    NaN                  NaN   
1165  722974582966214656                    NaN                  NaN   
1202  716439118184652801                    NaN                  NaN   
1228  713900603437621249                    NaN                  NaN   
1254  710658690886586372                    NaN                  NaN   
1274  709198395643068416                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1433  697463031882764288                    NaN                  NaN   
1598  686035780142297088           6.860340e+17         4.196984e+09   
1634  684225744407494656           6.842229e+17         4.196984e+09   
1635  684222868335505415                    NaN                  NaN   
1662  682962037429899265                    NaN                  NaN   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1779  677716515794329600                    NaN                  NaN   
1843  675853064436391936                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   

                      timestamp  \
313   2017-02-24 21:54:03 +0000   
342   2017-02-16 04:45:50 +0000   
433   2017-01-15 17:52:40 +0000   
516   2016-12-19 23:06:23 +0000   
784   2016-09-11 22:20:06 +0000   
902   2016-07-28 01:00:57 +0000   
1068  2016-06-08 02:41:38 +0000   
1120  2016-05-13 16:15:54 +0000   
1165  2016-04-21 02:25:47 +0000   
1202  2016-04-03 01:36:11 +0000   
1228  2016-03-27 01:29:02 +0000   
1254  2016-03-18 02:46:49 +0000   
1274  2016-03-14 02:04:08 +0000   
1351  2016-02-28 21:25:30 +0000   
1433  2016-02-10 16:51:59 +0000   
1598  2016-01-10 04:04:10 +0000   
1634  2016-01-05 04:11:44 +0000   
1635  2016-01-05 04:00:18 +0000   
1662  2016-01-01 16:30:13 +0000   
1663  2016-01-01 06:22:03 +0000   
1779  2015-12-18 05:06:23 +0000   
1843  2015-12-13 01:41:41 +0000   
2335  2015-11-16 16:11:11 +0000   

                                                 source  \
313   <a href="http://twitter.com/download/iphone" r...   
342   <a href="http://twitter.com/download/iphone" r...   
433   <a href="http://twitter.com/download/iphone" r...   
516   <a href="http://twitter.com/download/iphone" r...   
784   <a href="http://twitter.com/download/iphone" r...   
902   <a href="http://twitter.com/download/iphone" r...   
1068  <a href="http://twitter.com/download/iphone" r...   
1120  <a href="http://twitter.com/download/iphone" r...   
1165  <a href="http://twitter.com/download/iphone" r...   
1202  <a href="http://twitter.com/download/iphone" r...   
1228  <a href="http://twitter.com/download/iphone" r...   
1254  <a href="http://twitter.com/download/iphone" r...   
1274  <a href="http://twitter.com/download/iphone" r...   
1351  <a href="http://twitter.com/download/iphone" r...   
1433  <a href="http://twitter.com/download/iphone" r...   
1598  <a href="http://twitter.com/download/iphone" r...   
1634  <a href="http://twitter.com/download/iphone" r...   
1635  <a href="http://twitter.com/download/iphone" r...   
1662  <a href="http://twitter.com/download/iphone" r...   
1663  <a href="http://twitter.com/download/iphone" r...   
1779  <a href="http://twitter.com/download/iphone" r...   
1843  <a href="http://twitter.com/download/iphone" r...   
2335  <a href="http://twitter.com/download/iphone" r...   

                             

##### There are  23 rows where rating_denominator is not equal to 10

##### Lets take a closer look at the rating_numerator

In [31]:
tweet_archive_data.rating_numerator.value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
420       2
0         2
15        2
75        2
80        1
20        1
24        1
26        1
44        1
50        1
60        1
165       1
84        1
88        1
144       1
182       1
143       1
666       1
960       1
1776      1
17        1
27        1
45        1
99        1
121       1
204       1
Name: rating_numerator, dtype: int64

#####  According to WeRateDogs,dog ratings almost always have a denominator of 10. The numerators, though? Almost always greater than 10. 11/10, 12/10, 13/10, etc. Why? Because ["they're good dogs Brent."](https://knowyourmeme.com/memes/theyre-good-dogs-brent) WeRateDogs has over 4 million followers and has received international media coverage.

In [32]:
tweet_archive_data.query("rating_numerator > 14").sort_values('rating_numerator')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
285   838916489579200512                    NaN                  NaN   
291   838085839343206401           8.380855e+17         2.894131e+09   
55    881633300179243008           8.816070e+17         4.738443e+07   
1663  682808988178739200           6.827884e+17         4.196984e+09   
516   810984652412424192                    NaN                  NaN   
1712  680494726643068929                    NaN                  NaN   
763   778027034220126208                    NaN                  NaN   
1433  697463031882764288                    NaN                  NaN   
1274  709198395643068416                    NaN                  NaN   
1202  716439118184652801                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
340   832215909146226688                    NaN                  NaN   
695   786709082849828864                    NaN                  NaN   
1254  710658690886586372                    NaN                  NaN   
433   820690176645140481                    NaN                  NaN   
1843  675853064436391936                    NaN                  NaN   
1228  713900603437621249                    NaN                  NaN   
1635  684222868335505415                    NaN                  NaN   
1634  684225744407494656           6.842229e+17         4.196984e+09   
1779  677716515794329600                    NaN                  NaN   
902   758467244762497024                    NaN                  NaN   
290   838150277551247360           8.381455e+17         2.195506e+07   
1120  731156023742988288                    NaN                  NaN   
2074  670842764863651840                    NaN                  NaN   
188   855862651834028034           8.558616e+17         1.943518e+08   
189   855860136149123072           8.558585e+17         1.361572e+07   
313   835246439529840640           8.352460e+17         2.625958e+07   
979   749981277374128128                    NaN                  NaN   

                      timestamp  \
285   2017-03-07 00:57:32 +0000   
291   2017-03-04 17:56:49 +0000   
55    2017-07-02 21:58:53 +0000   
1663  2016-01-01 06:22:03 +0000   
516   2016-12-19 23:06:23 +0000   
1712  2015-12-25 21:06:00 +0000   
763   2016-09-20 00:24:34 +0000   
1433  2016-02-10 16:51:59 +0000   
1274  2016-03-14 02:04:08 +0000   
1202  2016-04-03 01:36:11 +0000   
1351  2016-02-28 21:25:30 +0000   
340   2017-02-16 13:11:49 +0000   
695   2016-10-13 23:23:56 +0000   
1254  2016-03-18 02:46:49 +0000   
433   2017-01-15 17:52:40 +0000   
1843  2015-12-13 01:41:41 +0000   
1228  2016-03-27 01:29:02 +0000   
1635  2016-01-05 04:00:18 +0000   
1634  2016-01-05 04:11:44 +0000   
1779  2015-12-18 05:06:23 +0000   
902   2016-07-28 01:00:57 +0000   
290   2017-03-04 22:12:52 +0000   
1120  2016-05-13 16:15:54 +0000   
2074  2015-11-29 05:52:33 +0000   
188   2017-04-22 19:15:32 +0000   
189   2017-04-22 19:05:32 +0000   
313   2017-02-24 21:54:03 +0000   
979   2016-07-04 15:00:45 +0000   

                                                 source  \
285   <a href="http://twitter.com/download/iphone" r...   
291   <a href="http://twitter.com/download/iphone" r...   
55    <a href="http://twitter.com/download/iphone" r...   
1663  <a href="http://twitter.com/download/iphone" r...   
516   <a href="http://twitter.com/download/iphone" r...   
1712  <a href="http://twitter.com/download/iphone" r...   
763   <a href="http://twitter.com/download/iphone" r...   
1433  <a href="http://twitter.com/download/iphone" r...   
1274  <a href="http://twitter.com/download/iphone" r...   
1202  <a href="http://twitter.com/download/iphone" r...   
1351  <a href="http://twitter.com/download/iphone" r...   
340   <a href="http://twitter.com/download/iphone" r...   
695   <a href="http://twitter.com/download/iphone" r...   
1254  <a href="http://twitter.com/download/iphone" r...   
433   <a href="http://twit

##### There are 28 rows with rating_numerator greater than 14 ..... are these exceeding the ratings for "They are good dog brent" these may be erroneous.

##### Lets take a look at the rows where rating_numerator is greater than 14 and rating_denominator is not equal to 10

In [33]:
tweet_archive_data.query("rating_numerator > 14 & rating_denominator != 10").sort_values('rating_numerator')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1663  682808988178739200           6.827884e+17         4.196984e+09   
516   810984652412424192                    NaN                  NaN   
1433  697463031882764288                    NaN                  NaN   
1274  709198395643068416                    NaN                  NaN   
1202  716439118184652801                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1254  710658690886586372                    NaN                  NaN   
433   820690176645140481                    NaN                  NaN   
1843  675853064436391936                    NaN                  NaN   
1228  713900603437621249                    NaN                  NaN   
1635  684222868335505415                    NaN                  NaN   
1634  684225744407494656           6.842229e+17         4.196984e+09   
1779  677716515794329600                    NaN                  NaN   
902   758467244762497024                    NaN                  NaN   
1120  731156023742988288                    NaN                  NaN   
313   835246439529840640           8.352460e+17         2.625958e+07   

                      timestamp  \
1663  2016-01-01 06:22:03 +0000   
516   2016-12-19 23:06:23 +0000   
1433  2016-02-10 16:51:59 +0000   
1274  2016-03-14 02:04:08 +0000   
1202  2016-04-03 01:36:11 +0000   
1351  2016-02-28 21:25:30 +0000   
1254  2016-03-18 02:46:49 +0000   
433   2017-01-15 17:52:40 +0000   
1843  2015-12-13 01:41:41 +0000   
1228  2016-03-27 01:29:02 +0000   
1635  2016-01-05 04:00:18 +0000   
1634  2016-01-05 04:11:44 +0000   
1779  2015-12-18 05:06:23 +0000   
902   2016-07-28 01:00:57 +0000   
1120  2016-05-13 16:15:54 +0000   
313   2017-02-24 21:54:03 +0000   

                                                 source  \
1663  <a href="http://twitter.com/download/iphone" r...   
516   <a href="http://twitter.com/download/iphone" r...   
1433  <a href="http://twitter.com/download/iphone" r...   
1274  <a href="http://twitter.com/download/iphone" r...   
1202  <a href="http://twitter.com/download/iphone" r...   
1351  <a href="http://twitter.com/download/iphone" r...   
1254  <a href="http://twitter.com/download/iphone" r...   
433   <a href="http://twitter.com/download/iphone" r...   
1843  <a href="http://twitter.com/download/iphone" r...   
1228  <a href="http://twitter.com/download/iphone" r...   
1635  <a href="http://twitter.com/download/iphone" r...   
1634  <a href="http://twitter.com/download/iphone" r...   
1779  <a href="http://twitter.com/download/iphone" r...   
902   <a href="http://twitter.com/download/iphone" r...   
1120  <a href="http://twitter.com/download/iphone" r...   
313   <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1663  I'm aware that I could've said 20/16, but here...                  NaN   
516   Meet Sam. She smiles 24/7 &amp; secretly aspir...                  NaN   
1433  Happy Wednesday here's a bucket of pups. 44/40...                  NaN   
1274  From left to right:\nCletus, Jerome, Alejandro...                  NaN   
1202  This is Bluebert. He just saw that both #Final...                  NaN   
1351  Here is a whole flock of puppers.  60/50 I'll ...                  NaN   
1254  Here's a brigade of puppers. All look very pre...                  NaN   
433   The floofs have been released I repeat the flo...                  NaN   
1843  Here we have an entire platoon of puppers. Tot...                  NaN   
1228  Happy Saturday here's 9 puppers on a bench. 99...                  NaN   
1635  Someone help the girl is being mugged. Several...                  NaN   
1634  Two sneaky puppers were not initially seen, mo...                  NaN   
1779  IT'S PUPPERGEDDON. Total of 144/120 ...I think...                  NaN   
902   Why does this never happen at my front door......                

#####  There are 16 records that have ratings_numerator greater than 14 and ratings_denominator not equal to 10
##### These records definitely look errorneous as the ratings do not make any sense

##### Let's take a look at the last four columns of tweet_archive_data dataframe

In [34]:
tweet_archive_data[['doggo','floofer','pupper','puppo']].head(10)

doggo floofer pupper puppo
0   None    None   None  None
1   None    None   None  None
2   None    None   None  None
3   None    None   None  None
4   None    None   None  None
5   None    None   None  None
6   None    None   None  None
7   None    None   None  None
8   None    None   None  None
9  doggo    None   None  None

##### Looks like these four columns represent data for dog breed , which can be represented in one column called dog_breed

### Visual Assessment
#### For tweet_image_data

In [75]:
tweet_image_data

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

##### Based on just visual assessmet we can see that there are 12 observations in this table and it holds the images url for the images in each tweet and the dog breed predictions using the image recognition software. It looks like this table is pretty clean, no NaN values found so far. Although the column names are not descriptive.

### Programmatic Assessment
#### For tweet_image_date

In [76]:
tweet_image_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


###### There are 12 columns and 2075 observations
###### Column names are not very descriptive.
######  None of the columns have NaN values

###### lets see if there are any duplicate tweet_ids

In [37]:
tweet_image_data[tweet_image_data.tweet_id.duplicated()]

Empty DataFrame
Columns: [tweet_id, jpg_url, img_num, p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog]
Index: []

###### There are no duplicate tweet_ids so these are all unique tweets or retweets may be
###### Lets see if there are duplicates in image urls, i.e if a image is used my multiple tweets.

In [38]:
tweet_image_data[tweet_image_data.jpg_url.duplicated(keep=False)].sort_values('jpg_url')

tweet_id                                            jpg_url  \
480   675354435921575936  https://pbs.twimg.com/ext_tw_video_thumb/67535...   
1297  752309394570878976  https://pbs.twimg.com/ext_tw_video_thumb/67535...   
1864  842892208864923648  https://pbs.twimg.com/ext_tw_video_thumb/80710...   
1641  807106840509214720  https://pbs.twimg.com/ext_tw_video_thumb/80710...   
1703  817181837579653120  https://pbs.twimg.com/ext_tw_video_thumb/81596...   
1691  815966073409433600  https://pbs.twimg.com/ext_tw_video_thumb/81596...   
1705  817423860136083457  https://pbs.twimg.com/ext_tw_video_thumb/81742...   
1858  841833993020538882  https://pbs.twimg.com/ext_tw_video_thumb/81742...   
1715  819004803107983360    https://pbs.twimg.com/media/C12whDoVEAALRxa.jpg   
1718  819015337530290176    https://pbs.twimg.com/media/C12whDoVEAALRxa.jpg   
1716  819006400881917954    https://pbs.twimg.com/media/C12x-JTVIAAzdfl.jpg   
1717  819015331746349057    https://pbs.twimg.com/media/C12x-JTVIAAzdfl.jpg   
1746  823269594223824897    https://pbs.twimg.com/media/C2kzTGxWEAEOpPL.jpg   
1738  822244816520155136    https://pbs.twimg.com/media/C2kzTGxWEAEOpPL.jpg   
1740  822489057087389700    https://pbs.twimg.com/media/C2oRbOuWEAAbVSl.jpg   
1742  822647212903690241    https://pbs.twimg.com/media/C2oRbOuWEAAbVSl.jpg   
1767  826958653328592898    https://pbs.twimg.com/media/C3nygbBWQAAjwcW.jpg   
1789  829878982036299777    https://pbs.twimg.com/media/C3nygbBWQAAjwcW.jpg   
1785  829374341691346946    https://pbs.twimg.com/media/C4KHj-nWQAA3poV.jpg   
1903  851953902622658560    https://pbs.twimg.com/media/C4KHj-nWQAA3poV.jpg   
2041  885311592912609280    https://pbs.twimg.com/media/C4bTH6nWMAAX_bJ.jpg   
1791  830583320585068544    https://pbs.twimg.com/media/C4bTH6nWMAAX_bJ.jpg   
224   670319130621435904    https://pbs.twimg.com/media/CU1zsMSUAAAS0qW.jpg   
1345  759159934323924993    https://pbs.twimg.com/media/CU1zsMSUAAAS0qW.jpg   
1590  798644042770751489    https://pbs.twimg.com/media/CU3mITUWIAAfyQS.jpg   
241   670444955656130560    https://pbs.twimg.com/media/CU3mITUWIAAfyQS.jpg   
1589  798628517273620480    https://pbs.twimg.com/media/CUN4Or5UAAAa5K4.jpg   
85    667509364010450944    https://pbs.twimg.com/media/CUN4Or5UAAAa5K4.jpg   
327   671896809300709376    https://pbs.twimg.com/media/CVMOlMiWwAA4Yxl.jpg   
1591  798665375516884993    https://pbs.twimg.com/media/CVMOlMiWwAA4Yxl.jpg   
...                  ...                                                ...   
1571  794983741416415232    https://pbs.twimg.com/media/CvT6IV6WEAQhhV5.jpg   
1530  789530877013393408    https://pbs.twimg.com/media/CvT6IV6WEAQhhV5.jpg   
1538  790723298204217344    https://pbs.twimg.com/media/CvaYgDOWgAEfjls.jpg   
1533  789986466051088384    https://pbs.twimg.com/media/CvaYgDOWgAEfjls.jpg   
1683  813944609378369540    https://pbs.twimg.com/media/Cveg1-NXgAASaaT.jpg   
1534  790277117346975746    https://pbs.twimg.com/media/Cveg1-NXgAASaaT.jpg   
1539  790946055508652032    https://pbs.twimg.com/media/CvoBPWRWgAA4het.jpg   
1699  816829038950027264    https://pbs.twimg.com/media/CvoBPWRWgAA4het.jpg   
1544  791672322847637504    https://pbs.twimg.com/media/CvyVxQRWEAAdSZS.jpg   
1564  793614319594401792    https://pbs.twimg.com/media/CvyVxQRWEAAdSZS.jpg   
1804  832215726631055365    https://pbs.twimg.com/media/CwJR1okWIAA6XMp.jpg   
1561  793286476301799424    https://pbs.twimg.com/media/CwJR1okWIAA6XMp.jpg   
1566  793962221541933056    https://pbs.twimg.com/media/CwS4aqZXUAAe3IO.jpg   
1652  809808892968534016    https://pbs.twimg.com/media/CwS4aqZXUAAe3IO.jpg   
1572  795076730285391872    https://pbs.twimg.com/media/CwiuEJmW8AAZnit.jpg   
1755  824796380199809024    https://pbs.twimg.com/media/CwiuEJmW8AAZnit.jpg   
1579  796177847564038144    https://pbs.twimg.com/media/Cwx99rpW8AMk_Ie.jpg   
1578  796149749086875649    https://pbs.twimg.com/media/Cwx99rpW8AMk_Ie.jpg   
1610  801167903437357056    https://pbs.twimg.com/media/Cx5R8wPVEA

###### Interesting to see image url is shared by more than one tweet, what does this mean?
###### Since image predictions data is created for the twitter ids in tweet_archive_data, lets take a look at the tweet archive data for a sample of these tweets sharing the same image url.

In [39]:
tweet_archive_data.query("tweet_id in (675354435921575936,752309394570878976)")

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
949   752309394570878976                    NaN                  NaN   
1865  675354435921575936                    NaN                  NaN   

                      timestamp  \
949   2016-07-11 01:11:51 +0000   
1865  2015-12-11 16:40:19 +0000   

                                                 source  \
949   <a href="http://twitter.com/download/iphone" r...   
1865  <a href="http://twitter.com" rel="nofollow">Tw...   

                                                   text  retweeted_status_id  \
949   RT @dog_rates: Everyone needs to watch this. 1...         6.753544e+17   
1865  Everyone needs to watch this. 13/10 https://t....                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
949               4.196984e+09  2015-12-11 16:40:19 +0000   
1865                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
949   https://twitter.com/dog_rates/status/675354435...                13   
1865  https://twitter.com/dog_rates/status/675354435...                13   

      rating_denominator  name doggo floofer pupper puppo  
949                   10  None  None    None   None  None  
1865                  10  None  None    None   None  None

In [40]:
tweet_archive_data.query("tweet_id in (711694788429553666,761371037149827077)")

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
872   761371037149827077                    NaN                  NaN   
1246  711694788429553666                    NaN                  NaN   

                      timestamp  \
872   2016-08-05 01:19:35 +0000   
1246  2016-03-20 23:23:54 +0000   

                                                 source  \
872   <a href="http://twitter.com/download/iphone" r...   
1246  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
872   RT @dog_rates: Oh. My. God. 13/10 magical af h...         7.116948e+17   
1246  Oh. My. God. 13/10 magical af https://t.co/Ezu...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
872               4.196984e+09  2016-03-20 23:23:54 +0000   
1246                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
872   https://twitter.com/dog_rates/status/711694788...                13   
1246  https://twitter.com/dog_rates/status/711694788...                13   

      rating_denominator  name doggo floofer pupper puppo  
872                   10  None  None    None   None  None  
1246                  10  None  None    None   None  None

###### If you look at the retweeted_status_id , it holds the original tweet_id if it is a retweet. This is one way to recognize retweets. 

### Visual Assessment
#### For tweet_count_data

In [77]:
tweet_count_data

tweet_id  retweet_count  favorite_count
0     892420643555336193           8257           37848
1     892177421306343426           6101           32523
2     891815181378084864           4039           24483
3     891689557279858688           8400           41202
4     891327558926688256           9106           39378
5     891087950875897856           3022           19795
6     890971913173991426           1999           11570
7     890729181411237888          18341           63867
8     890609185150312448           4157           27212
9     890240255349198849           7172           31193
10    890006608113172480           7126           29985
11    889880896479866881           4836           27196
12    889665388333682689           9759           47035
13    889638837579907072           4400           26532
14    889531135344209921           2187           14782
15    889278841981685760           5212           24671
16    888917238123831296           4376           28477
17    888804989199671297           4162           24991
18    888554962724278272           3442           19384
19    888078434458587136           3394           21271
20    887705289381826560           5240           29513
21    887517139158093824          11370           45240
22    887473957103951883          17666           67438
23    887343217045368832          10135           32927
24    887101392804085760           5798           29898
25    886983233522544640           7556           34356
26    886736880519319552           3185           11797
27    886680336477933568           4342           21944
28    886366144734445568           3111           20728
29    886267009285017600              4             116
...                  ...            ...             ...
2310  666411507551481857            322             434
2311  666407126856765440             40             105
2312  666396247373291520             83             162
2313  666373753744588802             89             184
2314  666362758909284353            560             760
2315  666353288456101888             71             215
2316  666345417576210432            132             287
2317  666337882303524864             90             192
2318  666293911632134144            344             489
2319  666287406224695296             63             142
2320  666273097616637952             76             170
2321  666268910803644416             35             101
2322  666104133288665088           6384           14028
2323  666102155909144576             11              78
2324  666099513787052032             66             151
2325  666094000022159362             72             161
2326  666082916733198337             44             113
2327  666073100786774016            159             316
2328  666071193221509120             58             142
2329  666063827256086533            213             465
2330  666058600524156928             57             109
2331  666057090499244032            138             290
2332  666055525042405380            235             428
2333  666051853826850816            831            1196
2334  666050758794694657             57             130
2335  666049248165822465             42             106
2336  666044226329800704            136             292
2337  666033412701032449             43             122
2338  666029285002620928             46             125
2339  666020888022790149            498            2528

[2340 rows x 3 columns]

##### Based on visual assement we can see that this data set consists for three observations tweet_id, retweet_count and favorite_count
##### Looks like there aren't any NaN values

### Programmatic Assessment
#### For tweet_count_data

In [78]:
tweet_count_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2340 entries, 0 to 2339
Data columns (total 3 columns):
tweet_id          2340 non-null int64
retweet_count     2340 non-null int64
favorite_count    2340 non-null int64
dtypes: int64(3)
memory usage: 54.9 KB


###### There are 2340 observations and three columns. There are no NaN values

## Quality

#### twitter-archive-enhanced.csv

###### 1. in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id , retweeted_status_user_id, retweeted_status_timestamp columns have majority null values.
###### 2. retweeted_status_id  is populated for retweets. To prepare a simple dataset to join with tweet counts lets remove the retweets.
###### 3. in_reply_to_status_id is populated for tweet replies.Lets remove these tweet replies.
###### 4. name column has majority of the rows ( 745) have None populated.There are also 109 rows that have invalid values for name with values like ( 'the','a','an','such' etc).
###### 5. Invalid values in ratings numerator and rating denominator. There are 16 records that have ratings numerator greater than 14 and ratings denominator not equal to 10.
###### 6. timestamp column is not in date type.
###### 7. When one of the columns in columns doggo, floofer, pupper, puppo is populated, other columns are populated with a value "None" . This should be represented as a Null value.
###### 8. Source column data is not very useful for the analysis. Delete this column.

#### image-predictions.tsv

###### 9. column names must be renamed to something more descriptive about the data it holds
###### 10. Some breed name values populated in p1_dog are starting with upper case and some are starting with lower case letter.


### Tidiness 

#### twitter-archive-enhanced.csv

###### 1. Dog breed value can be restructured to be stored in one column instead of four different columns.

#### image-predictions.tsv

###### 2.  Re structure  or drop the columns that are unnecessary. such as img_num,p2, p2_conf,p2_dog,p3,p3_conf,p3_dog 

###### 3. twitter_archive_data has 2356 rows and twitter_image_data has 2075 rows. After removing the retweets we need to merge the data in these datasets with tweet_counts to have the data in single data frame



# Clean

#### Define
1. Select all the rows that are retweets
2. Delete all rows that are retweets
3. select the rows that are tweet replies
4. Delete the rows that are tweet replies
3. Delete columns in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id , retweeted_status_user_id, retweeted_status_timestamp after removing the retweets

#### Code

In [79]:
# Create a copy of the data frame
# we do not want to loose the original data set
tweet_archive_data_clean = tweet_archive_data.copy()


In [80]:
tweet_archive_data_clean.head(1)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None

In [81]:
# creating boolean series True for retweeted_status_id isnull 
# original tweets have retweeted_status_id as null
retweet_is_null = pd.isnull(tweet_archive_data_clean['retweeted_status_id'])
tweet_archive_data_clean = tweet_archive_data_clean[retweet_is_null]

In [82]:
# creating boolean series True for reply_status_id isnull 
# original tweets have reply_status_id as null
reply_is_null = pd.isnull(tweet_archive_data_clean['in_reply_to_status_id'])
tweet_archive_data_clean = tweet_archive_data_clean[reply_is_null]

In [83]:
# Drop columns in_reply_to_status_id,in_reply_to_user_id,retweeted_status_id,
# retweeted_status_user_id, retweeted_status_timestamp 

tweet_archive_data_clean.drop(['in_reply_to_status_id', 'in_reply_to_user_id',
                            'retweeted_status_id','retweeted_status_user_id',
                            'retweeted_status_timestamp'],inplace = True, axis=1)

#### Test

In [84]:
tweet_archive_data_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2097 non-null int64
timestamp             2097 non-null object
source                2097 non-null object
text                  2097 non-null object
expanded_urls         2094 non-null object
rating_numerator      2097 non-null int64
rating_denominator    2097 non-null int64
name                  2097 non-null object
doggo                 2097 non-null object
floofer               2097 non-null object
pupper                2097 non-null object
puppo                 2097 non-null object
dtypes: int64(3), object(9)
memory usage: 213.0+ KB


#### Define
name column has majority of the rows ( 745) have None populated.There are also 109 rows that have invalid values for name with values like ( 'the','a','an','such' etc).
1. Gather the values in the name column that start with lowercase. These seem to be invalid values. Store them in a list.
2. Replace these values with null values
3. Replace None with Null values


#### Code

In [86]:
# Create a boolean series for name starting with lowercase letter
name_lower_case = tweet_archive_data_clean['name'].str[0].str.islower()

In [87]:
import numpy as np

In [88]:
# store these name values into an array
invalid_names = tweet_archive_data_clean[name_lower_case].name.values

In [89]:
invalid_names

array(['such', 'a', 'quite', 'quite', 'not', 'one', 'incredibly', 'a', 'a',
       'very', 'my', 'one', 'not', 'his', 'one', 'a', 'a', 'a', 'an',
       'very', 'actually', 'a', 'just', 'getting', 'mad', 'very', 'this',
       'unacceptable', 'all', 'a', 'old', 'a', 'infuriating', 'a', 'a',
       'a', 'an', 'a', 'a', 'very', 'getting', 'just', 'a', 'the', 'the',
       'actually', 'by', 'a', 'officially', 'a', 'the', 'the', 'a', 'a',
       'a', 'a', 'life', 'a', 'one', 'a', 'a', 'a', 'light', 'just',
       'space', 'a', 'the', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a',
       'an', 'a', 'the', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a',
       'a', 'quite', 'a', 'an', 'a', 'an', 'the', 'the', 'a', 'a', 'an',
       'a', 'a', 'a', 'a'], dtype=object)

In [90]:
# replace the above values in the name column with null values
tweet_archive_data_clean.name.replace(invalid_names,np.nan, inplace = True)

In [91]:
# replace the "None" value name column with null values
tweet_archive_data_clean.name.replace('None',np.nan, inplace = True)

#### Test

In [93]:
tweet_archive_data_clean.name.unique()

array(['Phineas', 'Tilly', 'Archie', 'Darla', 'Franklin', nan, 'Jax',
       'Zoey', 'Cassie', 'Koda', 'Bruno', 'Ted', 'Stuart', 'Oliver', 'Jim',
       'Zeke', 'Ralphus', 'Gerald', 'Jeffrey', 'Canela', 'Maya', 'Mingus',
       'Derek', 'Roscoe', 'Waffles', 'Jimbo', 'Maisey', 'Earl', 'Lola',
       'Kevin', 'Yogi', 'Noah', 'Bella', 'Grizzwald', 'Rusty', 'Gus',
       'Stanley', 'Alfy', 'Koko', 'Rey', 'Gary', 'Elliot', 'Louis',
       'Jesse', 'Romeo', 'Bailey', 'Duddles', 'Jack', 'Steven', 'Beau',
       'Snoopy', 'Shadow', 'Emmy', 'Aja', 'Penny', 'Dante', 'Nelly',
       'Ginger', 'Benedict', 'Venti', 'Goose', 'Nugget', 'Cash', 'Jed',
       'Sebastian', 'Sierra', 'Monkey', 'Harry', 'Kody', 'Lassie', 'Rover',
       'Napolean', 'Boomer', 'Cody', 'Rumble', 'Clifford', 'Dewey',
       'Scout', 'Gizmo', 'Walter', 'Cooper', 'Harold', 'Shikha', 'Lili',
       'Jamesy', 'Coco', 'Sammy', 'Meatball', 'Paisley', 'Albus',
       'Neptune', 'Belle', 'Quinn', 'Zooey', 'Dave', 'Jersey', 'Hobbes',


#### Define
1. Change the timestamp column to date data type

#### Code

In [98]:
tweet_archive_data_clean.timestamp = pd.to_datetime( tweet_archive_data_clean.timestamp,infer_datetime_format=True)

In [99]:
tweet_archive_data_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2097 entries, 0 to 2355
Data columns (total 12 columns):
tweet_id              2097 non-null int64
timestamp             2097 non-null datetime64[ns]
source                2097 non-null object
text                  2097 non-null object
expanded_urls         2094 non-null object
rating_numerator      2097 non-null int64
rating_denominator    2097 non-null int64
name                  1390 non-null object
doggo                 2097 non-null object
floofer               2097 non-null object
pupper                2097 non-null object
puppo                 2097 non-null object
dtypes: datetime64[ns](1), int64(3), object(8)
memory usage: 293.0+ KB


#### Define
There are 23 rows where rating denominator is not equal to 10
 For consistency purposes I onyl want to keep the records where denominator is  equal to 10.

1. Filter out the rows where ratings denominator is not equal to 10

#### Code

In [100]:
tweet_archive_data_clean = tweet_archive_data_clean[tweet_archive_data_clean.rating_denominator == 10]

#### Test

In [101]:
tweet_archive_data_clean.rating_denominator.unique()

array([10])

#### Define
1. Filter out extreme outliers in Ratings numerator
i.e delete rows where ratings numerator is ( 1776,420,75) any ratings greater than equal to 74

In [102]:
tweet_archive_data_clean.rating_numerator.unique()

array([  13,   12,   14,    5,   11,    6,   10,    0,   75,   27,    3,
          7,    8,    9,    4, 1776,   26,    2,    1,  420])

#### Code

In [103]:
tweet_archive_data_clean = tweet_archive_data_clean[tweet_archive_data_clean.rating_numerator < 74]

#### Test

In [104]:
tweet_archive_data_clean.rating_numerator.unique()

array([13, 12, 14,  5, 11,  6, 10,  0, 27,  3,  7,  8,  9,  4, 26,  2,  1])

#### Define
Delete source and expanded_urls column

#### Code

In [105]:
tweet_archive_data_clean.drop(['source','expanded_urls'],inplace = True, axis=1)

#### Test

In [106]:
tweet_archive_data_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2077 entries, 0 to 2355
Data columns (total 10 columns):
tweet_id              2077 non-null int64
timestamp             2077 non-null datetime64[ns]
text                  2077 non-null object
rating_numerator      2077 non-null int64
rating_denominator    2077 non-null int64
name                  1385 non-null object
doggo                 2077 non-null object
floofer               2077 non-null object
pupper                2077 non-null object
puppo                 2077 non-null object
dtypes: datetime64[ns](1), int64(3), object(6)
memory usage: 178.5+ KB


##### Tidiness for twitter_archive_data

#### Define

1. Create a new columns called breed_type and populate it using the data in the columns doggo,floofer,pupper and puppo. 


#### Code

In [109]:
tweet_archive_data_clean[['doggo', 'floofer','pupper', 'puppo']].sample(5)

doggo floofer  pupper puppo
796    None    None    None  None
1524   None    None    None  None
1409   None    None    None  None
985   doggo    None    None  None
1327   None    None  pupper  None

In [110]:
# replace None value with Null
columns = ['doggo', 'floofer','pupper', 'puppo']
for col in columns:
    tweet_archive_data_clean[col].replace('None',np.nan, inplace = True)

In [152]:
tweet_archive_data_clean.query('tweet_id in (855851453814013952,892420643555336193,890240255349198849)')

tweet_id           timestamp  \
0    892420643555336193 2017-08-01 16:23:56   
9    890240255349198849 2017-07-26 15:59:51   
191  855851453814013952 2017-04-22 18:31:02   

                                                  text  rating_numerator  \
0    This is Phineas. He's a mystical boy. Only eve...                13   
9    This is Cassie. She is a college pup. Studying...                14   
191  Here's a puppo participating in the #ScienceMa...                13   

     rating_denominator     name  doggo floofer pupper  puppo  
0                    10  Phineas    NaN     NaN    NaN    NaN  
9                    10   Cassie  doggo     NaN    NaN    NaN  
191                  10      NaN  doggo     NaN    NaN  puppo

In [174]:
def breed_type(in_cols):
    breed = []
    for col in in_cols:
        if col in in_cols:
            breed.append(col)
    if len(breed)>0:
        return ','.join(breed)
    else:
        return None
    
    

In [176]:
test_df=tweet_archive_data_clean.query('tweet_id in (855851453814013952,892420643555336193,890240255349198849)')

In [177]:
test_df[columns].apply(breed_type,axis=1 )

0             None
9            doggo
191    doggo,puppo
dtype: object

In [178]:
tweet_archive_data_clean['breed'] = tweet_archive_data_clean[columns].apply(breed_type,axis=1 )

In [179]:
tweet_archive_data_clean[['breed','doggo', 'floofer','pupper', 'puppo']]

breed  doggo floofer  pupper  puppo
0       None    NaN     NaN     NaN    NaN
1       None    NaN     NaN     NaN    NaN
2       None    NaN     NaN     NaN    NaN
3       None    NaN     NaN     NaN    NaN
4       None    NaN     NaN     NaN    NaN
5       None    NaN     NaN     NaN    NaN
6       None    NaN     NaN     NaN    NaN
7       None    NaN     NaN     NaN    NaN
8       None    NaN     NaN     NaN    NaN
9      doggo  doggo     NaN     NaN    NaN
10      None    NaN     NaN     NaN    NaN
11      None    NaN     NaN     NaN    NaN
12     puppo    NaN     NaN     NaN  puppo
13      None    NaN     NaN     NaN    NaN
14     puppo    NaN     NaN     NaN  puppo
15      None    NaN     NaN     NaN    NaN
16      None    NaN     NaN     NaN    NaN
17      None    NaN     NaN     NaN    NaN
18      None    NaN     NaN     NaN    NaN
20      None    NaN     NaN     NaN    NaN
21      None    NaN     NaN     NaN    NaN
22      None    NaN     NaN     NaN    NaN
23      None    NaN     NaN     NaN    NaN
24      None    NaN     NaN     NaN    NaN
25      None    NaN     NaN     NaN    NaN
26      None    NaN     NaN     NaN    NaN
27      None    NaN     NaN     NaN    NaN
28      None    NaN     NaN     NaN    NaN
29    pupper    NaN     NaN  pupper    NaN
31      None    NaN     NaN     NaN    NaN
...      ...    ...     ...     ...    ...
2325    None   None    None    None   None
2326    None   None    None    None   None
2327    None   None    None    None   None
2328    None   None    None    None   None
2329    None   None    None    None   None
2330    None   None    None    None   None
2331    None   None    None    None   None
2332    None   None    None    None   None
2333    None   None    None    None   None
2334    None   None    None    None   None
2336    None   None    None    None   None
2337    None   None    None    None   None
2338    None   None    None    None   None
2339    None   None    None    None   None
2340    None   None    None    None   None
2341    None   None    None    None   None
2342    None   None    None    None   None
2343    None   None    None    None   None
2344    None   None    None    None   None
2345    None   None    None    None   None
2346    None   None    None    None   None
2347    None   None    None    None   None
2348    None   None    None    None   None
2349    None   None    None    None   None
2350    None   None    None    None   None
2351    None   None    None    None   None
2352    None   None    None    None   None
2353    None   None    None    None   None
2354    None   None    None    None   None
2355    None   None    None    None   None

[2077 rows x 5 columns]

In [180]:
tweet_archive_data_clean.drop(['doggo', 'floofer', 'pupper', 'puppo'], 
    inplace = True, axis = 1)

In [181]:
tweet_archive_data_clean.breed.unique()

array([None, 'doggo', 'puppo', 'pupper', 'floofer', 'doggo,puppo',
       'doggo,floofer', 'doggo,pupper'], dtype=object)

#### Test

In [182]:
tweet_archive_data_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2077 entries, 0 to 2355
Data columns (total 7 columns):
tweet_id              2077 non-null int64
timestamp             2077 non-null datetime64[ns]
text                  2077 non-null object
rating_numerator      2077 non-null int64
rating_denominator    2077 non-null int64
name                  1385 non-null object
breed                 336 non-null object
dtypes: datetime64[ns](1), int64(3), object(3)
memory usage: 129.8+ KB


#### Quality for  tweet_image_data

#### Define
1. Drop the columns img_num,p1_dog,p2, p2_conf, p2_dog, p3, p3_conf,p3_dog
2. Rename the columns jpg_url, p1, p1_conf and p1_dog to descriptive column names

#### Code

In [183]:
# Create a copy of the data frame
# we do not want to update the original data set
tweet_image_data_clean = tweet_image_data.copy()



In [184]:
tweet_image_data_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [185]:
tweet_image_data_clean.drop(['img_num','p1_dog','p2', 'p2_conf', 'p2_dog', 'p3', 
                  'p3_conf', 'p3_dog'], inplace = True, axis = 1)

In [186]:
tweet_image_data_clean.rename(columns = {'jpg_url': 'image_url', 
                               'p1': 'prediction', 
                               'p1_conf': 'confidence_level'}, inplace = True)

In [187]:
tweet_image_data_clean.head()

tweet_id                                        image_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

               prediction  confidence_level  
0  Welsh_springer_spaniel          0.465074  
1                 redbone          0.506826  
2         German_shepherd          0.596461  
3     Rhodesian_ridgeback          0.408143  
4      miniature_pinscher          0.560311

#### Tidiness
Join the three data sets into one single data frame

In [188]:
tweet_archive_data_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2077 entries, 0 to 2355
Data columns (total 7 columns):
tweet_id              2077 non-null int64
timestamp             2077 non-null datetime64[ns]
text                  2077 non-null object
rating_numerator      2077 non-null int64
rating_denominator    2077 non-null int64
name                  1385 non-null object
breed                 336 non-null object
dtypes: datetime64[ns](1), int64(3), object(3)
memory usage: 129.8+ KB


In [189]:
tweet_image_data_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 4 columns):
tweet_id            2075 non-null int64
image_url           2075 non-null object
prediction          2075 non-null object
confidence_level    2075 non-null float64
dtypes: float64(1), int64(1), object(2)
memory usage: 64.9+ KB


In [190]:
tweet_count_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2340 entries, 0 to 2339
Data columns (total 3 columns):
tweet_id          2340 non-null int64
retweet_count     2340 non-null int64
favorite_count    2340 non-null int64
dtypes: int64(3)
memory usage: 54.9 KB


#### Code

In [191]:
# Merge the three data frames
twitter_archive_master = tweet_archive_data_clean.merge(tweet_count_data,
                                                   on = 'tweet_id', how = 'inner')
twitter_archive_master = twitter_archive_master.merge(tweet_image_data_clean, 
                                            on = 'tweet_id', how = 'left')

##### Test

In [192]:
twitter_archive_master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id              2075 non-null int64
timestamp             2075 non-null datetime64[ns]
text                  2075 non-null object
rating_numerator      2075 non-null int64
rating_denominator    2075 non-null int64
name                  1383 non-null object
breed                 336 non-null object
retweet_count         2075 non-null int64
favorite_count        2075 non-null int64
image_url             1949 non-null object
prediction            1949 non-null object
confidence_level      1949 non-null float64
dtypes: datetime64[ns](1), float64(1), int64(5), object(5)
memory usage: 210.7+ KB


In [193]:
twitter_archive_master.query("tweet_id in ('890729181411237888','890240255349198849')")

tweet_id           timestamp  \
7  890729181411237888 2017-07-28 00:22:40   
9  890240255349198849 2017-07-26 15:59:51   

                                                text  rating_numerator  \
7  When you watch your owner call another dog a g...                13   
9  This is Cassie. She is a college pup. Studying...                14   

   rating_denominator    name  breed  retweet_count  favorite_count  \
7                  10     NaN   None          18341           63867   
9                  10  Cassie  doggo           7172           31193   

                                         image_url  prediction  \
7  https://pbs.twimg.com/media/DFyBahAVwAAhUTd.jpg  Pomeranian   
9  https://pbs.twimg.com/media/DFrEyVuW0AAO3t9.jpg    Pembroke   

   confidence_level  
7          0.566142  
9          0.511319

In [194]:
tweet_image_data_clean.query("tweet_id in ('890729181411237888','890240255349198849')")

tweet_id                                        image_url  \
2065  890240255349198849  https://pbs.twimg.com/media/DFrEyVuW0AAO3t9.jpg   
2067  890729181411237888  https://pbs.twimg.com/media/DFyBahAVwAAhUTd.jpg   

      prediction  confidence_level  
2065    Pembroke          0.511319  
2067  Pomeranian          0.566142

In [195]:
tweet_count_data.query("tweet_id in ('890729181411237888','890240255349198849')")

tweet_id  retweet_count  favorite_count
7  890729181411237888          18341           63867
9  890240255349198849           7172           31193

# Store 

In [196]:
# Store tables to files
tweet_count_data.to_csv('tweet_count_data_clean.csv', encoding='utf-8', index=False)
tweet_image_data_clean.to_csv('tweet_image_data_clean.csv', encoding='utf-8', index=False)
tweet_archive_data_clean.to_csv('tweet_archive_data_clean.csv', encoding='utf-8', index=False)
twitter_archive_master.to_csv('twitter_archive_master.csv', encoding='utf-8', index=False)

### Code for data Analysis and Visualization is in act_report.ipynb